File: binary_tree/same_tree.rb
# 100
# Time: O(N) Space: O(1)
# A DnC approach
def is_same_tree(p, q)
  # recursive base case
  return true if p.nil? && q.nil?

  # DnC base problem
  return false if p.nil? && q || q.nil? && p || p.val != q.val

  # return solution, true if left and right child trees are the same, false otherwise
  left_same && right_same
end

File: binary_tree/same_tree.js


var isSameTree = function(p, q) {
  if (p === null && q === null) {
    return true;
  } else if (p === null || q === null || p.val != q.val) {
    return false;
  }

  let leftSame = isSameTree(p.left, q.left);
  let rightSame = isSameTree(p.right, q.right);

  return leftSame && rightSame;
};
